api/lib/domain/usecases/get-external-authentication-redirection-url.js

```diff
@@ -15,12 +17,58 @@ module.exports = async function getExternalAuthenticationRedirectionUrl({
   const user = await userRepository.getBySamlId(externalUser.samlId);
 
   if (user) {
-    const token = tokenService.createAccessTokenForSaml(user.id);
-    await userRepository.updateLastLoggedAt({ userId: user.id });
-    return `/?token=${encodeURIComponent(token)}&user-id=${user.id}`;
-  } else {
-    const externalUserToken = tokenService.createIdTokenForUserReconciliation(externalUser);
-
-    return `/campagnes?externalUser=${encodeURIComponent(externalUserToken)}`;
+    return await _getUrlWithAccessToken({
+      user,
+      externalUser,
+      tokenService,
+      userRepository,
+      authenticationMethodRepository,
+    });
   }
+
+  return _getUrlForReconciliationPage({ tokenService, externalUser });
 };
+
+async function _getUrlWithAccessToken({
+  user,
+  externalUser,
+  tokenService,
+  userRepository,
+  authenticationMethodRepository,
+}) {
+  const token = tokenService.createAccessTokenForSaml(user.id);
+  await userRepository.updateLastLoggedAt({ userId: user.id });
+  await _saveUserFirstAndLastName(authenticationMethodRepository, user, externalUser);
+  return `/?token=${encodeURIComponent(token)}&user-id=${user.id}`;
+}
+
+async function _saveUserFirstAndLastName(authenticationMethodRepository, user, externalUser) {
+  const authenticationMethod = await authenticationMethodRepository.findOneByUserIdAndIdentityProvider({
```

Comment:
**Remarque** : En testant j'avais noté ceci 
"De mon coté il m'a mis à jour mon authentication-methods même si je n'ai pas changé les informations"

A priori la cause serait ce `findOneByUserIdAndIdentityProvider` qui retourne une `authenticationComplement` undefined.

La méthode privée `externalUserFirstAndLastNameMatchesAuthenticationMethodFirstAndLastName` ne fait alors pas le match ici `authenticationMethod.authenticationComplement`

Reply:
Bien vu. Effectivement je reproduis le problème : `updatedAt` est mis à jour même sur le nom et le prénom envoyés par le GAR sont identiques à ceux en base. Par contre, je ne m'explique pas pourquoi le retour de `findOneByUserIdAndIdentityProvider` serait `undefined` ? On devrait en trouver une pour cet `userId` et 'GAR'…

Reply:
j'ai l'impression qu'il y a un soucis dans la façon dont on a stocké ces infos dans l'authentication complement coté GAR.
Il ne reconnait pas l'objet ? 🤔

Reply:
Il manque des tests pour cette méthode coté repo, on aurait du tester et vérifier que pour PE on récupère tel infos coté `authenticationComplement`, PIX autre chose et GAR rien...
Du coup si tu veux compléter 😇

Reply:
Merci @AndreiaPena pour ce travail d'enquête ! 

Du coup, j'ai ajouté trois commits pour :
- ajouter les cas de tests manquants
- corriger un autre problème au passage (une erreur masquée)
- corriger le problème dans le cas du GAR (on aurait du le faire dans la PR précédente)